lib/HTTP/Message.pm

```diff
@@ -402,6 +402,20 @@ sub decoded_content
 		}
 	    }
 	}
+	elsif ($self->content_is_json) {
+	    my $charset = lc(
+	        $opt{charset} ||
+		$opt{default_charset} ||
+		$self->content_charset ||
```

Comment:
that is interesting that you choose to not include `$self->content_type_charset`, which specifically uses the `charset` from the `Content-Type` header (like `application/json; charset=utf-16`). I'd thought that that would be the preferred way.

Also, it is discouraged to send the BOM, according to [RFC 7158 – JSON §8.1 String and Character Issues / Character Encoding](https://tools.ietf.org/html/rfc7158#section-8.1)

So, I'd suggest to certainly add `content_type_charset` before `content_charset`

Reply:
Sorry, my reading of what you wrote in #90 was that you'd want to ignore the charset header; if we're happy to use it (I certainly am!), then we could have it take the same code flow as text/* which is easier. Updated.

Reply:
if you want `decoded_content` do something smart, and extend "textual context" beyond `text/*` type (like we apparently also seem to do for `application/xml` ... then use what has been specified in the `Content-Type`

What I meant, was that this behaviour is outside the specs, only `text/*` have `charset` ... and thus should be ignored and should be passed in to the processor (that would be the module parsing the content-body into Perl structure). It is 'binary data' and 'we' shouldn't try to be smart but leave it to the processor. Just ask yourself, why munge on binary data for json or xml and not on audio-samples ...

but okay ... let's be 'smart' - but that is not my preference